README.rst
Flask-JWT
==============

.. image:: https://img.shields.io/travis/mattupstate/flask-jwt.svg
    :target: https://travis-ci.org/mattupstate/flask-jwt
    :alt: Latest Build

.. image:: https://img.shields.io/pypi/v/flask-jwt.svg
    :target: https://pypi.python.org/pypi/Flask-JWT/
    :alt: Latest Version

.. image:: https://img.shields.io/coveralls/mattupstate/flask-jwt.svg
    :target: https://coveralls.io/r/mattupstate/flask-jwt

.. image:: https://img.shields.io/pypi/dm/flask-jwt.svg
    :target: https://pypi.python.org/pypi//Flask-JWT/
    :alt: Downloads

.. image:: https://img.shields.io/pypi/l/flask-jwt.svg
    :target: https://pypi.python.org/pypi/Flask-JWT/
    :alt: License

JWT (JSON Web Tokens) for Flask applications


Resources
---------

- `Documentation <http://packages.python.org/Flask-JWT/>`_
- `Issue Tracker <http://github.com/mattupstate/flask-jwt/issues>`_
- `Code <http://github.com/mattupstate/flask-jwt/>`_
- `Development Version
  <http://github.com/mattupstate/flask-jwt/zipball/develop#egg=Flask-JWTy-dev>`_

Tests
-----

To run tests::

    python ./setup.py test


CHANGES
Flask-JWT Changelog
===================

Here you can see the full list of changes between each Flask-JWT release.

Version 0.3.3
-------------

Released March 7th 2018 (by AbdealiJK)

- Fixed an error when finding attribute 'id'

Version 0.3.2
-------------

Released November 3rd 2015

- Fixed an Authorization header conditional bug

Version 0.3.1
-------------

Released October 26th 2015

- Fix a bug with `auth_request_handler`
- Deprecate `auth_request_handler`

Version 0.3.0
-------------

Released October 15th 2015

.. note:: This release includes many breaking changes

- Fix major implementation issue with encoding/decoding tokens
- Changed new configuration options to align with PyJWT
- Changed `current_user` to `current_identity`

Version 0.2.0
-------------

Released June 10th 2014

- Fixed an issue where `current_user` was not None
- Added a response handler hook to be able to adjust auth response(s)
- Removed the configurable handlers in favor of decorators
- Removed pyjwt dependency


Version 0.1.0
-------------

Released March 5th 2014

- Initial release


tests.py

from unittest import TestCase

import flask_jwt
from flask import Flask, json

from nose.tools import *


class User(object):
    def __init__(self, **kwargs):
        for k, v in kwargs.iteritems():
            setattr(self, k, v)


def create_app():
    app = Flask(__name__)
    app.config['SECRET_KEY'] = 'super-secret'
    app.config['JWT_AUTH_URL_RULE'] = '/auth'
    app.config['JWT_AUTH_ENDPOINT'] = 'jwt_auth'
    jwt = flask_jwt.JWT(app)

    @jwt.authentication_handler
    def authenticate(username, password):
        if username == 'joe' and password == 'pass':
            return User(id=1, username='joe')
        None

    client = app.test_client()
    return app, client



class JWTTestCase(TestCase):

    def test_initialize(self):
        app = Flask(__name__)
        app.config['SECRET_KEY'] = 'super-secret'
        jwt = flask_jwt.JWT(app)
        assert_true(isinstance(jwt, flask_jwt.JWT))
        assert_equals(1, len(app.url_map._rules))

    def test_adds_auth_endpoint(self):
        app = Flask(__name__)
        app.config['SECRET_KEY'] = 'super-secret'
        app.config['JWT_AUTH_URL_RULE'] = '/auth'
        app.config['JWT_AUTH_ENDPOINT'] = 'jwt_auth'
        flask_jwt.JWT(app)
        rules = [str(r) for r in app.url_map._rules]
        assert_true('/auth' in rules)

    def test_auth_endpoint_with_valid_request(self):
        app, client = create_app()
        r = client.post('/auth',
            headers={'content-type': 'application/json'},
            data=json.dumps({
                'username': 'joe',
                'password': 'pass'
            }))
        assert_equals(200, r.status_code)
        jdata = json.loads(r.data)
        assert_true('token' in jdata)

    def test_auth_endpoint_with_invalid_request(self):
        app, client = create_app()
        # Invalid request
        r = client.post('/auth',
            headers={'content-type': 'application/json'},
            data=json.dumps({
                'username': 'joe'
            }))
        assert_equals(400, r.status_code)
        jdata = json.loads(r.data)
        assert_true('error' in jdata)
        assert_equals(jdata['error'], 'Bad Request')
        assert_true('description' in jdata)
        assert_equals(jdata['description'], 'Missing required credentials')
        assert_true('status_code' in jdata)
        assert_equals(jdata['status_code'], 400)

    def test_auth_endpoint_with_invalid_credentials(self):
        app, client = create_app()
        # Invalid credentials
        r = client.post('/auth',
            headers={'content-type': 'application/json'},
            data=json.dumps({
                'username': 'bogus',
                'password': 'bogus'
            }))
        assert_equals(400, r.status_code)
        jdata = json.loads(r.data)
        assert_true('error' in jdata)
        assert_equals(jdata['error'], 'Bad Request')
        assert_true('description' in jdata)
        assert_equals(jdata['description'], 'Invalid credentials')
        assert_true('status_code' in jdata)
        assert_equals(jdata['status_code'], 400)

    def test_jwt_required_decorator_with_valid_token(self):
        app, client = create_app()

        @app.route('/protected')
        @flask_jwt.jwt_required()
        def protected():
            return 'success'

        r = client.post('/auth',
            headers={'content-type': 'application/json'},
            data=json.dumps({
                'username': 'joe',
                'password': 'pass'
            }))

        jdata = json.loads(r.data)
        token = jdata['token']

        r = client.get('/protected',
            headers={'authorization': 'Bearer ' + token})
        assert_equals(200, r.status_code)
        assert_equals('success', r.data)


    def test_jwt_required_decorator_with_missing_token(self):
        app, client = create_app()

        @app.route('/protected')
        @flask_jwt.jwt_required()
        def protected():
            return 'success'

        r = client.get('/protected')
        jdata = json.loads(r.data)
        assert_equals(401, r.status_code)
        assert_equals(401, jdata['status_code'])
        assert_equals('Authorization Required', jdata['error'])
        assert_equals('Authorization header was missing', jdata['description'])
        assert_equals('JWT realm="Login Required"', r.headers['WWW-Authenticate'])


tests/conftest.py
# -*- coding: utf-8 -*-
"""
    tests.conftest
    ~~~~~~~~~~~~~~

    Test fixtures and what not
"""

import logging

import pytest

from flask import Flask
from datetime import datetime, timedelta

import flask_jwt

logging.basicConfig(level=logging.DEBUG)


class User(object):
    def __init__(self, id, username, password, role=None):
        self.id = id
        self.username = username
        self.password = password
        if role:
            self.role = role

    def __str__(self):
        return "User(id='%s')" % self.id


@pytest.fixture(scope='function')
def jwt():
    return flask_jwt.JWT()


@pytest.fixture(scope='function')
def user():
    return User(id=1, username='joe', password='pass')


@pytest.fixture(scope='function')
def user_with_role():
    return User(id=2, username='jane', password='pass', role='user')


@pytest.fixture(scope='function')
def user_with_roles():
    return User(id=3, username='alice', password='pass', role=['user', 'foo', 'bar'])


@pytest.fixture(scope='function')
def app(jwt, user):
    app = Flask(__name__)
    app.debug = True
    app.config['SECRET_KEY'] = 'super-secret'

    @jwt.authentication_handler
    def authenticate(username, password):
        if username == user.username and password == user.password:
            return user
        return None

    @jwt.identity_handler
    def load_user(payload):
        if payload['identity'] == user.id:
            return user

    jwt.init_app(app)

    @app.route('/protected')
    @flask_jwt.jwt_required()
    def protected():
        return 'success'

    return app


@pytest.fixture(scope='function')
def app_with_role(jwt, user, user_with_role, user_with_roles):
    app = Flask(__name__)
    app.debug = True
    app.config['SECRET_KEY'] = 'super-secret'
    users = [user, user_with_role, user_with_roles]

    @jwt.authentication_handler
    def authenticate(username, password):
        for u in users:
            if username == u.username and password == u.password:
                return u
        return None

    @jwt.identity_handler
    def load_user(payload):
        for u in users:
            if payload['identity'] == u.id:
                return u

    @jwt.payload_handler
    def make_payload(identity):
        iat = datetime.utcnow()
        exp = iat + timedelta(seconds=300)
        nbf = iat
        id = getattr(identity, 'id')
        try:
            role = getattr(identity, 'role')
            return {'exp': exp, 'iat': iat, 'nbf': nbf, 'identity': id, 'role': role}
        except AttributeError:
            return {'exp': exp, 'iat': iat, 'nbf': nbf, 'identity': id}

    jwt.init_app(app)

    @app.route('/protected')
    @flask_jwt.jwt_required()
    def protected():
        return 'success'

    @app.route('/role/protected/admin')
    @flask_jwt.jwt_required(roles='admin')
    def admin_protected():
        return 'success'

    @app.route('/role/protected/multi')
    @flask_jwt.jwt_required(roles=['admin', 'user'])
    def admin_user_protected():
        return 'success'

    @app.route('/role/protected/user')
    @flask_jwt.jwt_required(roles='user')
    def user_protected():
        return 'success'

    return app


@pytest.fixture(scope='function')
def app_with_role_trust_jwt(jwt, user, user_with_role, user_with_roles):
    app = Flask(__name__)
    app.debug = True
    app.config['SECRET_KEY'] = 'super-secret'
    app.config['JWT_ROLE'] = 'my_role'
    users = [user, user_with_role, user_with_roles]

    @jwt.authentication_handler
    def authenticate(username, password):
        for u in users:
            if username == u.username and password == u.password:
                return u
        return None

    @jwt.identity_handler
    def load_user(payload):
        return payload

    @jwt.payload_handler
    def make_payload(identity):
        iat = datetime.utcnow()
        exp = iat + timedelta(seconds=300)
        nbf = iat
        id = getattr(identity, 'id')
        try:
            role = getattr(identity, 'role')
            return {'exp': exp, 'iat': iat, 'nbf': nbf, 'identity': id, 'my_role': role}
        except AttributeError:
            return {'exp': exp, 'iat': iat, 'nbf': nbf, 'identity': id}

    jwt.init_app(app)

    @app.route('/protected')
    @flask_jwt.jwt_required()
    def protected():
        return 'success'

    @app.route('/role/protected/user')
    @flask_jwt.jwt_required(roles='user')
    def user_protected():
        return 'success'

    @app.route('/role/protected/multi')
    @flask_jwt.jwt_required(roles=['admin', 'user'])
    def admin_user_protected():
        return 'success'

    @app.route('/role/protected/admin')
    @flask_jwt.jwt_required(roles='admin')
    def admin_protected():
        return 'success'

    return app


@pytest.fixture(scope='function')
def client(app):
    return app.test_client()


@pytest.fixture(scope='function')
def dictuserapp(jwt, dictuser):
    app = Flask(__name__)
    app.debug = True
    app.config['SECRET_KEY'] = 'super-secret'

    @jwt.authentication_handler
    def authenticate(username, password):
        if username == dictuser['username'] and password == dictuser['password']:
            return dictuser
        return None

    @jwt.identity_handler
    def load_user(payload):
        if payload['identity'] == dictuser['id']:
            return dictuser

    jwt.init_app(app)

    @app.route('/protected')
    @flask_jwt.jwt_required()
    def protected():
        return 'success'

    return app


@pytest.fixture(scope='function')
def dictuser():
    return {'id': 1, 'username': 'joe', 'password': 'pass'}


flask_jwt.py

from datetime import datetime, timedelta
from functools import wraps

import jwt

from flask import current_app, request, jsonify, _request_ctx_stack
from flask.views import MethodView
from werkzeug.local import LocalProxy

current_user = LocalProxy(lambda: _request_ctx_stack.top.current_user)

_jwt = LocalProxy(lambda: current_app.extensions['jwt'])


def _default_payload_handler(user):
    return {
        'user_id': user.id,
        'exp': datetime.utcnow() + current_app.config['JWT_EXPIRATION_DELTA']
    }


def _default_encode_handler(payload):
    return jwt.encode(
        payload,
        current_app.config['JWT_SECRET_KEY'],
        current_app.config['JWT_ALGORITHM']
    ).decode('utf-8')


def _default_decode_handler(token):
    return jwt.decode(
        token,
        current_app.config['JWT_SECRET_KEY'],
        current_app.config['JWT_VERIFY'],
        current_app.config['JWT_VERIFY_EXPIRATION'],
        current_app.config['JWT_LEEWAY']
    )

CONFIG_DEFAULTS = {
    'JWT_DEFAULT_REALM': 'Login Required',
    'JWT_AUTH_URL_RULE': None,
    'JWT_AUTH_ENDPOINT': None,
    'JWT_ENCODE_HANDLER': _default_encode_handler,
    'JWT_DECODE_HANDLER': _default_decode_handler,
    'JWT_PAYLOAD_HANDLER': _default_payload_handler,
    'JWT_ALGORITHM': 'HS256',
    'JWT_VERIFY': True,
    'JWT_VERIFY_EXPIRATION': True,
    'JWT_LEEWAY': 0,
    'JWT_EXPIRATION_DELTA': timedelta(seconds=300)
}


def jwt_required(realm=None):
    def wrapper(fn):
        @wraps(fn)
        def decorator(*args, **kwargs):
            verify_jwt(realm)
            return fn(*args, **kwargs)
        return decorator
    return wrapper


class JWTError(Exception):
    def __init__(self, message, description, status_code=400, headers=None):
        self.message = message
        self.description = description
        self.status_code = status_code
        self.headers = headers


def verify_jwt(realm=None):
    realm = realm or current_app.config['JWT_DEFAULT_REALM']
    auth = request.headers.get('Authorization', None)

    if auth is None:
        raise JWTError('Authorization Required', 'Authorization header was missing', 401, {
            'WWW-Authenticate': 'JWT realm="%s"' % realm
        })

    parts = auth.split()

    if parts[0].lower() != 'bearer':
        return None
    elif len(parts) == 1:
        raise JWTError('Invalid JWT header', 'Token missing')
    elif len(parts) > 2:
        raise JWTError('Invalid JWT header', 'Token contains spaces')

    try:
        handler = current_app.config['JWT_DECODE_HANDLER']
        payload = handler(parts[1])
    except jwt.ExpiredSignature:
        raise JWTError('Invalid JWT', 'Token is expired')
    except jwt.DecodeError:
        raise JWTError('Invalid JWT', 'Token is undecipherable')

    _request_ctx_stack.top.current_user = _jwt.user_handler(payload)


class JWTAuthView(MethodView):

    def post(self):
        data = request.get_json()
        username = data.get('username', None)
        password = data.get('password', None)
        criterion = [username, password, len(data) == 2]

        if not all(criterion):
            raise JWTError('Bad Request', 'Missing required credentials', status_code=400)

        user = _jwt.authentication_callback(username=username, password=password)

        if user:
            payload_handler = current_app.config['JWT_PAYLOAD_HANDLER']
            payload = payload_handler(user)
            encode_handler = current_app.config['JWT_ENCODE_HANDLER']
            return jsonify({'token': encode_handler(payload)})
        else:
            raise JWTError('Bad Request', 'Invalid credentials')


class JWT(object):

    def __init__(self, app=None):
        if app is not None:
            self.app = app
            self.init_app(app)
        else:
            self.app = None

    def init_app(self, app):
        for k, v in CONFIG_DEFAULTS.iteritems():
            app.config.setdefault(k, v)
        app.config.setdefault('JWT_SECRET_KEY', app.config['SECRET_KEY'])

        if app.config['JWT_SECRET_KEY'] is None:
            raise RuntimeError('You must specify a SECRET_KEY or '
                               'JWT_SECRET_KEY configuration option')

        url_rule = app.config.get('JWT_AUTH_URL_RULE', None)
        endpoint = app.config.get('JWT_AUTH_ENDPOINT', None)

        if url_rule and endpoint:
            auth_view = JWTAuthView.as_view(app.config['JWT_AUTH_ENDPOINT'])
            app.add_url_rule(url_rule, methods=['POST'], view_func=auth_view)

        app.errorhandler(JWTError)(self._on_jwt_error)

        if not hasattr(app, 'extensions'):
            app.extensions = {}
        app.extensions['jwt'] = self

    def _on_jwt_error(self, e):
        callback = getattr(self, 'error_callback', None)
        if callback:
            return self.callback(e)
        else:
            return jsonify({
                'error': e.message,
                'description': e.description,
                'status_code': e.status_code
            }), e.status_code, e.headers

    def authentication_handler(self, callback):
        self.authentication_callback = callback
        return callback

    def user_handler(self, callback):
        self.user_callback = callback
        return callback

    def error_handler(self, callback):
        self.error_callback = callback
        return callback


example/app.py
from flask import Flask
from flask_jwt import JWT, jwt_required, current_identity
from werkzeug.security import safe_str_cmp

class User(object):
    def __init__(self, id, username, password):
        self.id = id
        self.username = username
        self.password = password

    def __str__(self):
        return "User(id='%s')" % self.id

users = [
    User(1, 'user1', 'abcxyz'),
    User(2, 'user2', 'abcxyz'),
]

username_table = {u.username: u for u in users}
userid_table = {u.id: u for u in users}

def authenticate(username, password):
    user = username_table.get(username, None)
    if user and safe_str_cmp(user.password.encode('utf-8'), password.encode('utf-8')):
        return user

def identity(payload):
    user_id = payload['identity']
    return userid_table.get(user_id, None)

app = Flask(__name__)
app.debug = True
app.config['SECRET_KEY'] = 'super-secret'

jwt = JWT(app, authenticate, identity)

@app.route('/protected')
@jwt_required()
def protected():
    return '%s' % current_identity

if __name__ == '__main__':
    app.run()


scripts/release.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
    make-release
    ~~~~~~~~~~~~

    Helper script that performs a release.  Does pretty much everything
    automatically for us.

    :copyright: (c) 2011 by Armin Ronacher.
    :license: BSD, see LICENSE for more details.
"""
import sys
import os
import re
from datetime import datetime, date
from subprocess import Popen, PIPE

_date_clean_re = re.compile(r'(\d+)(st|nd|rd|th)')


def installed_libraries():
    return Popen(['pip', 'freeze'], stdout=PIPE).communicate()[0]


def has_library_installed(library):
    return library + '==' in installed_libraries()


def parse_changelog():
    with open('CHANGES') as f:
        lineiter = iter(f)
        for line in lineiter:
            match = re.search('^Version\s+(.*)', line.strip())

            if match is None:
                continue

            version = match.group(1).strip()

            if lineiter.next().count('-') != len(line.strip()):
                fail('Invalid hyphen count below version line: %s', line.strip())

            while 1:
                released = lineiter.next().strip()
                if released:
                    break

            match = re.search(r'Released (\w+\s+\d+\w+\s+\d+)', released)

            if match is None:
                fail('Could not find release date in version %s' % version)

            datestr = parse_date(match.group(1).strip())

            return version, datestr


def bump_version(version):
    try:
        parts = map(int, version.split('.'))
    except ValueError:
        fail('Current version is not numeric')
    parts[-1] += 1
    return '.'.join(map(str, parts))


def parse_date(string):
    string = _date_clean_re.sub(r'\1', string)
    return datetime.strptime(string, '%B %d %Y')


def set_filename_version(filename, version_number, pattern):
    changed = []

    def inject_version(match):
        before, old, after = match.groups()
        changed.append(True)
        return before + version_number + after

    with open(filename) as f:
        contents = re.sub(r"^(\s*%s\s*=\s*')(.+?)(')(?sm)" % pattern,
                          inject_version, f.read())

    if not changed:
        fail('Could not find %s in %s', pattern, filename)

    with open(filename, 'w') as f:
        f.write(contents)


def set_init_version(version):
    info('Setting __init__.py version to %s', version)
    set_filename_version('flask_jwt/__init__.py', version, '__version__')


def set_setup_version(version):
    info('Setting setup.py version to %s', version)
    set_filename_version('setup.py', version, 'version')


def set_docs_version(version):
    info('Setting docs/conf.py version to %s', version)
    set_filename_version('docs/conf.py', version, 'version')


def build_and_upload():
    Popen([sys.executable, 'setup.py', 'sdist', 'build_sphinx', 'upload', 'upload_sphinx']).wait()


def fail(message, *args):
    print >> sys.stderr, 'Error:', message % args
    sys.exit(1)


def info(message, *args):
    print >> sys.stderr, message % args


def get_git_tags():
    return set(Popen(['git', 'tag'], stdout=PIPE).communicate()[0].splitlines())


def git_is_clean():
    return Popen(['git', 'diff', '--quiet']).wait() == 0


def make_git_commit(message, *args):
    message = message % args
    Popen(['git', 'commit', '-am', message]).wait()


def make_git_tag(tag):
    info('Tagging "%s"', tag)
    Popen(['git', 'tag', '-a', tag, '-m', '%s release' % tag]).wait()
    Popen(['git', 'push', '--tags']).wait()


def update_version(version):
    for f in [set_init_version, set_setup_version, set_docs_version]:
        f(version)


def get_branches():
    return set(Popen(['git', 'branch'], stdout=PIPE).communicate()[0].splitlines())


def branch_is(branch):
    return '* ' + branch in get_branches()


def main():
    os.chdir(os.path.join(os.path.dirname(__file__), '..'))

    rv = parse_changelog()

    if rv is None:
        fail('Could not parse changelog')

    version, release_date = rv

    tags = get_git_tags()

    for lib in ['Sphinx', 'Sphinx-PyPI-upload']:
        if not has_library_installed(lib):
            fail('Build requires that %s be installed', lib)

    if version in tags:
        fail('Version "%s" is already tagged', version)
    if release_date.date() != date.today():
        fail('Release date is not today')

    if not branch_is('master'):
        fail('You are not on the master branch')

    if not git_is_clean():
        fail('You have uncommitted changes in git')

    info('Releasing %s (release date %s)',
         version, release_date.strftime('%d/%m/%Y'))

    update_version(version)
    make_git_commit('Bump version number to %s', version)
    make_git_tag(version)
    build_and_upload()


if __name__ == '__main__':
    main()


flask_jwt/__init__.py
# -*- coding: utf-8 -*-
"""
    flask_jwt
    ~~~~~~~~~

    Flask-JWT module
"""

import logging
import warnings

from collections import OrderedDict
from datetime import datetime, timedelta
from functools import wraps

import jwt

from flask import current_app, request, jsonify, _request_ctx_stack
from werkzeug.local import LocalProxy

__version__ = '0.3.3'

logger = logging.getLogger(__name__)

current_identity = LocalProxy(lambda: getattr(_request_ctx_stack.top, 'current_identity', None))

_jwt = LocalProxy(lambda: current_app.extensions['jwt'])

CONFIG_DEFAULTS = {
    'JWT_DEFAULT_REALM': 'Login Required',
    'JWT_AUTH_URL_RULE': '/auth',
    'JWT_AUTH_ENDPOINT': 'jwt',
    'JWT_AUTH_USERNAME_KEY': 'username',
    'JWT_AUTH_PASSWORD_KEY': 'password',
    'JWT_ALGORITHM': 'HS256',
    'JWT_LEEWAY': timedelta(seconds=10),
    'JWT_AUTH_HEADER_PREFIX': 'JWT',
    'JWT_EXPIRATION_DELTA': timedelta(seconds=300),
    'JWT_NOT_BEFORE_DELTA': timedelta(seconds=0),
    'JWT_VERIFY_CLAIMS': ['signature', 'exp', 'nbf', 'iat'],
    'JWT_REQUIRED_CLAIMS': ['exp', 'iat', 'nbf']
}


def _default_jwt_headers_handler(identity):
    return None


def _default_jwt_payload_handler(identity):
    iat = datetime.utcnow()
    exp = iat + current_app.config.get('JWT_EXPIRATION_DELTA')
    nbf = iat + current_app.config.get('JWT_NOT_BEFORE_DELTA')
    if hasattr(identity, 'id'):
        identity = getattr(identity, 'id')
    else:
        identity = identity['id']
    return {'exp': exp, 'iat': iat, 'nbf': nbf, 'identity': identity}


def _default_jwt_encode_handler(identity):
    secret = current_app.config['JWT_SECRET_KEY']
    algorithm = current_app.config['JWT_ALGORITHM']
    required_claims = current_app.config['JWT_REQUIRED_CLAIMS']

    payload = _jwt.jwt_payload_callback(identity)
    missing_claims = list(set(required_claims) - set(payload.keys()))

    if missing_claims:
        raise RuntimeError('Payload is missing required claims: %s' % ', '.join(missing_claims))

    headers = _jwt.jwt_headers_callback(identity)

    return jwt.encode(payload, secret, algorithm=algorithm, headers=headers)


def _default_jwt_decode_handler(token):
    secret = current_app.config['JWT_SECRET_KEY']
    algorithm = current_app.config['JWT_ALGORITHM']
    leeway = current_app.config['JWT_LEEWAY']

    verify_claims = current_app.config['JWT_VERIFY_CLAIMS']
    required_claims = current_app.config['JWT_REQUIRED_CLAIMS']

    options = {
        'verify_' + claim: True
        for claim in verify_claims
    }

    options.update({
        'require_' + claim: True
        for claim in required_claims
    })

    return jwt.decode(token, secret, options=options, algorithms=[algorithm], leeway=leeway)


def _default_request_handler():
    auth_header_value = request.headers.get('Authorization', None)
    auth_header_prefix = current_app.config['JWT_AUTH_HEADER_PREFIX']

    if not auth_header_value:
        return

    parts = auth_header_value.split()

    if parts[0].lower() != auth_header_prefix.lower():
        raise JWTError('Invalid JWT header', 'Unsupported authorization type')
    elif len(parts) == 1:
        raise JWTError('Invalid JWT header', 'Token missing')
    elif len(parts) > 2:
        raise JWTError('Invalid JWT header', 'Token contains spaces')

    return parts[1]


def _default_auth_request_handler():
    data = request.get_json()
    username = data.get(current_app.config.get('JWT_AUTH_USERNAME_KEY'), None)
    password = data.get(current_app.config.get('JWT_AUTH_PASSWORD_KEY'), None)
    criterion = [username, password, len(data) == 2]

    if not all(criterion):
        raise JWTError('Bad Request', 'Invalid credentials')

    identity = _jwt.authentication_callback(username, password)

    if identity:
        access_token = _jwt.jwt_encode_callback(identity)
        return _jwt.auth_response_callback(access_token, identity)
    else:
        raise JWTError('Bad Request', 'Invalid credentials')


def _default_auth_response_handler(access_token, identity):
    return jsonify({'access_token': access_token.decode('utf-8')})


def _default_jwt_error_handler(error):
    logger.error(error)
    return jsonify(OrderedDict([
        ('status_code', error.status_code),
        ('error', error.error),
        ('description', error.description),
    ])), error.status_code, error.headers


def _jwt_required(realm):
    """Does the actual work of verifying the JWT data in the current request.
    This is done automatically for you by `jwt_required()` but you could call it manually.
    Doing so would be useful in the context of optional JWT access in your APIs.

    :param realm: an optional realm
    """
    token = _jwt.request_callback()

    if token is None:
        raise JWTError('Authorization Required', 'Request does not contain an access token',
                       headers={'WWW-Authenticate': 'JWT realm="%s"' % realm})

    try:
        payload = _jwt.jwt_decode_callback(token)
    except jwt.InvalidTokenError as e:
        raise JWTError('Invalid token', str(e))

    _request_ctx_stack.top.current_identity = identity = _jwt.identity_callback(payload)

    if identity is None:
        raise JWTError('Invalid JWT', 'User does not exist')


def jwt_required(realm=None):
    """View decorator that requires a valid JWT token to be present in the request

    :param realm: an optional realm
    """
    def wrapper(fn):
        @wraps(fn)
        def decorator(*args, **kwargs):
            _jwt_required(realm or current_app.config['JWT_DEFAULT_REALM'])
            return fn(*args, **kwargs)
        return decorator
    return wrapper


class JWTError(Exception):
    def __init__(self, error, description, status_code=401, headers=None):
        self.error = error
        self.description = description
        self.status_code = status_code
        self.headers = headers

    def __repr__(self):
        return 'JWTError: %s' % self.error

    def __str__(self):
        return '%s. %s' % (self.error, self.description)


def encode_token():
    return _jwt.encode_callback(_jwt.header_callback(), _jwt.payload_callback())


class JWT(object):

    def __init__(self, app=None, authentication_handler=None, identity_handler=None):
        self.authentication_callback = authentication_handler
        self.identity_callback = identity_handler

        self.auth_response_callback = _default_auth_response_handler
        self.auth_request_callback = _default_auth_request_handler
        self.jwt_encode_callback = _default_jwt_encode_handler
        self.jwt_decode_callback = _default_jwt_decode_handler
        self.jwt_headers_callback = _default_jwt_headers_handler
        self.jwt_payload_callback = _default_jwt_payload_handler
        self.jwt_error_callback = _default_jwt_error_handler
        self.request_callback = _default_request_handler

        if app is not None:
            self.init_app(app)

    def init_app(self, app):
        for k, v in CONFIG_DEFAULTS.items():
            app.config.setdefault(k, v)
        app.config.setdefault('JWT_SECRET_KEY', app.config['SECRET_KEY'])

        auth_url_rule = app.config.get('JWT_AUTH_URL_RULE', None)

        if auth_url_rule:
            if self.auth_request_callback == _default_auth_request_handler:
                assert self.authentication_callback is not None, (
                    'an authentication_handler function must be defined when using the built in '
                    'authentication resource')

            auth_url_options = app.config.get('JWT_AUTH_URL_OPTIONS', {'methods': ['POST']})
            auth_url_options.setdefault('view_func', self.auth_request_callback)
            app.add_url_rule(auth_url_rule, **auth_url_options)

        app.errorhandler(JWTError)(self._jwt_error_callback)

        if not hasattr(app, 'extensions'):  # pragma: no cover
            app.extensions = {}

        app.extensions['jwt'] = self

    def _jwt_error_callback(self, error):
        return self.jwt_error_callback(error)

    def authentication_handler(self, callback):
        """Specifies the identity handler function. This function receives two positional
        arguments. The first being the username the second being the password. It should return an
        object representing an authenticated identity. Example::

            @jwt.authentication_handler
            def authenticate(username, password):
                user = User.query.filter(User.username == username).scalar()
                if bcrypt.check_password_hash(user.password, password):
                    return user

        :param callback: the identity handler function
        """
        self.authentication_callback = callback
        return callback

    def identity_handler(self, callback):
        """Specifies the identity handler function. This function receives one positional argument
        being the JWT payload. For example::

            @jwt.identity_handler
            def identify(payload):
                return User.query.filter(User.id == payload['identity']).scalar()

        :param callback: the identity handler function
        """
        self.identity_callback = callback
        return callback

    def jwt_error_handler(self, callback):
        """Specifies the error handler function. Example::

            @jwt.error_handler
            def error_handler(e):
                return "Something bad happened", 400

        :param callback: the error handler function
        """
        self.jwt_error_callback = callback
        return callback

    def auth_response_handler(self, callback):
        """Specifies the authentication response handler function.

        :param callable callback: the auth response handler function
        """
        self.auth_response_callback = callback
        return callback

    def auth_request_handler(self, callback):
        """Specifies the authentication response handler function.

        :param callable callback: the auth request handler function

        .. deprecated
        """
        warnings.warn("This handler is deprecated. The recommended approach to have control over "
                      "the authentication resource is to disable the built-in  resource by "
                      "setting JWT_AUTH_URL_RULE=None and registering your own authentication "
                      "resource directly on your application.", DeprecationWarning, stacklevel=2)
        self.auth_request_callback = callback
        return callback

    def request_handler(self, callback):
        """Specifieds the request handler function. This function returns a JWT from the current
        request.

        :param callable callback: the request handler function
        """
        self.request_callback = callback
        return callback

    def jwt_encode_handler(self, callback):
        """Specifies the encoding handler function. This function receives a payload and signs it.

        :param callable callback: the encoding handler function
        """
        self.jwt_encode_callback = callback
        return callback

    def jwt_decode_handler(self, callback):
        """Specifies the decoding handler function. This function receives a
        signed payload and decodes it.

        :param callable callback: the decoding handler function
        """
        self.jwt_decode_callback = callback
        return callback

    def jwt_payload_handler(self, callback):
        """Specifies the JWT payload handler function. This function receives the return value from
        the ``identity_handler`` function

        Example::

            @jwt.payload_handler
            def make_payload(identity):
                return {'user_id': identity.id}

        :param callable callback: the payload handler function
        """
        self.jwt_payload_callback = callback
        return callback

    def jwt_headers_handler(self, callback):
        """Specifies the JWT header handler function. This function receives the return value from
        the ``identity_handler`` function.

        Example::

            @jwt.payload_handler
            def make_payload(identity):
                return {'user_id': identity.id}

        :param callable callback: the payload handler function
        """
        self.jwt_headers_callback = callback
        return callback
